mathics/builtin/layout.py

```diff
@@ -0,0 +1,846 @@
+# -*- coding: utf-8 -*-
+
+"""
+High level layout and formatting.
+"""
+
+import re
+
+from typing import Optional
+
+
+from mathics.builtin.base import (
+    Builtin,
+    BinaryOperator,
+    Operator,
+)
+from mathics.builtin.box.inout import GridBox, RowBox, _BoxedString, to_boxes
+from mathics.builtin.comparison import expr_min
+from mathics.core.formatter import lookup_method as lookup_format_method
+from mathics.builtin.lists import list_boxes
+from mathics.builtin.makeboxes import MakeBoxes
+from mathics.builtin.options import options_to_rules
+from mathics.builtin.tensors import get_dimensions
+
+from mathics.core.atoms import (
+    Integer,
+    Real,
+    String,
+    StringFromPython,
+)
+
+from mathics.core.element import EvalMixin
+from mathics.core.expression import Expression, BoxError
+from mathics.core.list import ListExpression
+from mathics.core.symbols import (
+    Symbol,
+    SymbolFullForm,
+    SymbolList,
+)
+
+from mathics.core.systemsymbols import (
+    SymbolMakeBoxes,
+    SymbolRowBox,
+    SymbolRule,
+)
+
+MULTI_NEWLINE_RE = re.compile(r"\n{2,}")
+
+SymbolNumberForm = Symbol("System`NumberForm")
+SymbolSuperscriptBox = Symbol("System`SuperscriptBox")
+SymbolSubscriptBox = Symbol("System`SubscriptBox")
+
+
+class Row(Builtin):
+    """
+    <dl>
+    <dt>'Row[{$expr$, ...}]'
```

Comment:
I would appreciate it if you'd add some spaces, like two when nesting XML tags. Not doing this was yet another one of those bone-headed weirdnesses of the original code.

Reply:
This is a boring thing that could be automatized at some point. Also, as soon as I finish with this refactor, I would like to start refactoring the documentation using a more canonical system, like sphinx. So, I am not sure how useful is to do this now.

Reply:
> So, I am not sure how useful is to do this now.

This kind of tradeoff  has been pestering me from the beginning. The more we delay the more work we will have to do in conversion, _unless_ there is automation to do it. 

As you recall from the earliest days of getting involved I have been advocating for using something like sphinx. (It or something like that was around from the days that this code was written. It was boneheaded to not build on it but rather ignore it and write something from scratch). 

However we should not kid ourselves that replacing it even had we done this earlier is a lot of work. And in the meantime I do feel strongly that our docs should be useful and we should continue to improve them.

But right now, I think we agree that the priorities have to be improving performance. Sort of up there would also be increasing contributors,  involvement, and hooking into the community. 

 So for the homegrown documentation system, I hold my nose (and probably complain _a_ _lot_.

Reply:
A little side digression about 
>  increasing contributors

When you have a code base that calls all tree nodes "leaves", and describes "eval" methods as "apply", doesn't seem to understand what a list literal is, but writes some homegrown (and possibly wrong) custom caching and its own autodoc system, (among numerous other examples) you are not going to attract the kind of experienced programmers that are needed here. And that has been in the back of my mind.